GenderPayGap.WebUI/Controllers/Report/ReportReviewAndSubmitController.cs

```diff
@@ -0,0 +1,210 @@
+            ControllerHelper.ThrowIfUserDoesNotHavePermissionsForGivenOrganisation(User, dataRepository, organisationId);
+            ControllerHelper.ThrowIfReportingYearIsOutsideOfRange(reportingYear);
+
+            if (!draftReturnService.DraftReturnExistsAndIsComplete(organisationId, reportingYear))
+            {
+                string nextPageUrl = Url.Action("ReportOverview", "ReportOverview", new { encryptedOrganisationId = encryptedOrganisationId, reportingYear = reportingYear });
+                StatusMessageHelper.SetStatusMessage(Response, "This report is not ready to submit. Complete the remaining sections", nextPageUrl);
+                return LocalRedirect(nextPageUrl);
+            }
+
+            DraftReturn draftReturn = draftReturnService.GetDraftReturn(organisationId, reportingYear);
+            if (draftReturnService.DraftReturnWouldBeNewlyLateIfSubmittedNow(draftReturn))
+            {
+                return RedirectToAction("LateSubmissionGet", "ReportReviewAndSubmit",
+                    new {encryptedOrganisationId = encryptedOrganisationId, reportingYear = reportingYear});
+            }
+
+            User user = ControllerHelper.GetGpgUserFromAspNetUser(User, dataRepository);
+            Return newReturn = returnService.CreateAndSaveReturnFromDraftReturn(draftReturn, user, Url);
+
+            return RedirectToAction("ReportConfirmation", "ReportConfirmation",
+                new
+                {
+                    encryptedOrganisationId = encryptedOrganisationId,
+                    reportingYear = reportingYear,
+                    confirmationId = Encryption.EncryptQuerystring(newReturn.ReturnId.ToString())
+                });
+        }
+
+
+        [HttpGet("{encryptedOrganisationId}/reporting-year-{reportingYear}/report/late-submission")]
+        public IActionResult LateSubmissionGet(string encryptedOrganisationId, int reportingYear)
+        {
+            ControllerHelper.Throw404IfFeatureDisabled(FeatureFlag.NewReportingJourney);
+
+            long organisationId = ControllerHelper.DecryptOrganisationIdOrThrow404(encryptedOrganisationId);
+            ControllerHelper.ThrowIfUserAccountRetiredOrEmailNotVerified(User, dataRepository);
+            ControllerHelper.ThrowIfUserDoesNotHavePermissionsForGivenOrganisation(User, dataRepository, organisationId);
+            ControllerHelper.ThrowIfReportingYearIsOutsideOfRange(reportingYear);
+
+            if (!draftReturnService.DraftReturnExistsAndIsComplete(organisationId, reportingYear))
+            {
+                string nextPageUrl = Url.Action("ReportOverview", "ReportOverview", new { encryptedOrganisationId = encryptedOrganisationId, reportingYear = reportingYear });
+                StatusMessageHelper.SetStatusMessage(Response, "This report is not ready to submit. Complete the remaining sections", nextPageUrl);
+                return LocalRedirect(nextPageUrl);
+            }

```

Comment:
I've tried commonising it (and pushed it as a commit).

A couple of things:
* We can't pull the whole `if` into a method, because we `return` in the middle.
But we can pull out the block inside the `{ }`.
* I've tried commonising - the file ends up the same length overall, but each method is a bit shorter (which is nice).

I think I'm happy either way - what do you think?

Reply:
I quite like it with the commonised section - it makes the methods shorter and the naming means that it's clear what's happening (overall) in those three lines!

Reply:
Great - thanks for checking 😄